refactor: avoid span_data and  instrumentation_scope clone for every single span processor

taisho6339 opened the pull request:
Fixes #1125

## Changes

Changed `SpanProcessor::on_end()` signature to accept `&SpanData` and `&InstrumentationScope` by reference. This eliminates unnecessary `SpanData` clones when multiple `SpanProcessor`s are configured. Performance improvement scales with the number of processors. Processors that need to store the data asynchronously should clone it explicitly.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

lalitb commented:
Thanks for the PR @taisho6339. Could you check if we can use a pattern similar to the one in the Logs SDK? In that implementation, the `InstrumentationScope` is passed by reference from the Logger to the Processor, allowing the Processor to decide whether to clone it or pass the reference to the exporter. I believe this approach was chosen to ensure the synchronous pipeline remains lightweight by avoiding cloning (even Arc overhead). This may lead to cloning of InstrumentationScope in case of batch processor, but that should also be cheap if the instrumentation-name is static string (which is the common use-case), and there are no attributes.

taisho6339 committed:
refactor: changed `SpanProcessor::on_end()` signature to accept `&SpanData` and `&InstrumentationScope` by reference
This eliminates unnecessary `SpanData` clones when multiple `SpanProcessor`s are configured. Performance improvement scales with the number of processors.
Processors that need to store the data asynchronously should clone it explicitly.

taisho6339 committed:
cargo fmt

lalitb committed:
Merge branch 'main' into issue-1125

taisho6339 committed:
Merge branch 'main' into issue-1125
